app/models/solution.rb

```diff
@@ -189,8 +189,8 @@ def downloaded? = !!downloaded_at
 
   def completed? = !!completed_at
   def published? = !!published_at
-
   def iterated? = status != :started
+  def submitted? = iterated? || completed? || published? || submissions.exists?
```

Comment:
Could you reorder these lines with the two above so it's in order of: submitted?, iterated?, completed?, published?

Reply:
I purposefully didn't do that as the submitted? check (`submissions.exists?`) is the only one that runs a query. Having it this order short-circuits in many cases.

Reply:
```suggestion
  # These are ordered this way for db lookup efficiency
  def submitted? = iterated? || completed? || published? || submissions.exists?
```